### PR TITLE
feature(frontend): adding instrumentation metrics

### PR DESCRIPTION
**Pull request by xoscar**
# Changes

This PR adds a set of changes to support instrumentation metrics from the frontend (server side), it also upgrades the open telemetry dependencies to the latest version. 

## Merge Requirements

For new features contributions please make sure you have completed the following
essential items:

* [ ] `CHANGELOG.md` updated to document new feature additions
* [ ] Appropriate documentation updates in the [docs](../docs/) folder

Maintainers will not merge until the above have been completed. If you're unsure
which docs need to be changed ping the
[@open-telemetry/demo-approvers](https://github.com/orgs/open-telemetry/teams/demo-approvers).

### CODE REVIEW

**Comment by fatsheep9146** [comment]
please resolve the conflicts first. @xoscar

File: src/frontend/Dockerfile [hunk 1]
```diff
@@ -18,7 +18,7 @@ RUN npm run build
 
 FROM node:18-alpine AS runner
 WORKDIR /app
-RUN apk add --no-cache protoc
+RUN apk add --no-cache protobuf-dev protoc
```
Comment (fatsheep9146) [inline src/frontend/Dockerfile:21]: I think this change is duplicated in #631, #632 #633. I'm not sure, is it necessary to add this change to every pr?

File: src/frontend/utils/telemetry/Instrumentation.js [hunk 2]
```diff
@@ -1,10 +1,21 @@
...
+const { PeriodicExportingMetricReader } = require('@opentelemetry/sdk-metrics');
 
 const sdk = new opentelemetry.NodeSDK({
   traceExporter: new OTLPTraceExporter(),
-  instrumentations: [ getNodeAutoInstrumentations() ]
-})
+  instrumentations: [
+    getNodeAutoInstrumentations({
+      '@opentelemetry/instrumentation-fs': {
+        enabled: false,
```
Comment (puckpuck) [inline src/frontend/utils/telemetry/Instrumentation.js:12]: why are we disabling this auto-instrumentation?

**Review by puckpuck — CHANGES_REQUESTED** [review]

File: src/frontend/utils/telemetry/Instrumentation.js (hunk 2 above)
Comment (xoscar) [reply src/frontend/utils/telemetry/Instrumentation.js:12]: This is a great question, so I upgrade all open telemetry dependencies to support the metrics feature. Once I upgraded the node auto instrumentation and tried to execute the next JS app I kept getting an error. It seems that the auto instrumentation fs package is breaking the nodejs functionality. Here's a screenshot:

Screenshot 2022-12-14 at 9 20 10


@puckpuck we might have to create an issue for the auto instrumentation package, it started happening when I upgraded to 0.34.0

Comment (puckpuck) [reply src/frontend/utils/telemetry/Instrumentation.js:12]: I'm able to replicate the above scenario as well with the `fs` instrumentation enabled.

I did some digging and found a [related issue](https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1315) to this which hasn't been resolved yet.

I would rather we don't have code specifically to get around SDK version upgrade issues.

**Comment by cartersocha** [comment]
Hey @pichlermarc if you get a chance would you mind giving this a quick review?

**Review by pichlermarc — APPROVED** [review]
Sorry for taking so long to review this. :grimacing: 

Looks great! Just one comment about the temporality preference env var. :slightly_smiling_face: 
We may also need to update [Metric Feature Coverage by Service](https://github.com/open-telemetry/opentelemetry-demo/blob/main/docs/metric_service_features.md).

The fix for the `fs-instrumentation` may only come around in the beginning of next year. :slightly_frowning_face: It is up to the author & approvers, but it may be worth keeping the workaround for now to get this merged and create a issue to remove the workaround in a follow up once the fix is released.

**Comment by cartersocha** [comment]
Hey @xoscar can you clean this up? I think we're mostly ready

**Comment by xoscar** [comment]
Hey @cartersocha @puckpuck this one should be ready for its final review

**Review by austinlparker — APPROVED** [review]
lgtm but can you update the docs as well?

**Comment by xoscar** [comment]
Hey @austinlparker what docs update would you like to be added? or just the change log update?

**Comment by austinlparker** [comment]
The feature matrix for metrics needs to be updated, right?

**Review by julianocosta89 — CHANGES_REQUESTED** [review]
I'm not able to build this PR.
I'm getting the following error:

```
=> ERROR [builder 8/8] RUN npm run build                                                                                                             45.7s
------
 > [builder 8/8] RUN npm run build:
#0 0.706
#0 0.706 > frontend@0.1.0 build
#0 0.706 > next build
#0 0.706
#0 1.141 warn  - SWC minify release candidate enabled. https://nextjs.org/docs/messages/swc-minify-enabled
#0 1.157 Attention: Next.js now collects completely anonymous telemetry regarding usage.
#0 1.158 This information is used to shape Next.js' roadmap and prioritize features.
#0 1.158 You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
#0 1.158 https://nextjs.org/telemetry
#0 1.158
#0 1.309 info  - Linting and checking validity of types...
#0 4.289
#0 4.289 ./pages/_document.tsx
#0 4.289 46:11  Warning: Do not use `<title>` element with `<Head />` component from `next/document`. Titles should defined at the page-level using `<Head />` from `next/head` instead. See: https://nextjs.org/docs/messages/no-title-in-document-head  @next/next/no-title-in-document-head
#0 4.289
#0 4.289 ./components/CartDropdown/CartDropdown.tsx
#0 4.289 30:6  Warning: React Hook useEffect has a missing dependency: 'onClose'. Either include it or remove the dependency array. If 'onClose' changes too often, find the parent component that defines it and wrap that definition in useCallback.  react-hooks/exhaustive-deps
#0 4.289
#0 4.289 info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
#0 9.018 info  - Creating an optimized production build...
#0 9.619 Browserslist: caniuse-lite is outdated. Please run:
#0 9.619   npx browserslist@latest --update-db
#0 9.619   Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
#0 9.681 warn  - Found lockfile missing swc dependencies, patching...
#0 10.10 info  - Downloading WASM swc package...
#0 12.49 warn  - Lockfile was successfully patched, please run "npm install" to ensure @next/swc dependencies are downloaded
#0 14.33 info  - Using experimental wasm build of next-swc
#0 37.40 warn  - Attempted to load @next/swc-linux-x64-gnu, but it was not installed
#0 37.40 warn  - Attempted to load @next/swc-linux-x64-gnux32, but it was not installed
#0 37.40 warn  - Attempted to load @next/swc-linux-x64-musl, but it was not installed
#0 43.22 Browserslist: caniuse-lite is outdated. Please run:
#0 43.22   npx browserslist@latest --update-db
#0 43.22   Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
#0 44.98 panicked at 'The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Error { kind: Unsupported, message: "operation not supported on this platform" }) }', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.9.1/src/registry.rs:170:10
#0 44.98
#0 44.98 Stack:
#0 44.98
#0 44.98 Error
#0 44.98     at module.exports.__wbg_new_693216e109162396 (/app/node_modules/next/wasm/@next/swc-wasm-nodejs/wasm.js:202:17)
#0 44.98     at wasm://wasm/059aa236:wasm-function[5445]:0xf90917
#0 44.98     at wasm://wasm/059aa236:wasm-function[13700]:0x11eca7b
#0 44.98     at wasm://wasm/059aa236:wasm-function[9995]:0x1186adc
#0 44.98     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 44.98     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 44.98     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 44.98     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 44.98     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 44.98     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 44.98
#0 44.98
#0 45.01 panicked at 'The global thread pool has not been initialized.: ThreadPoolBuildError { kind: GlobalPoolAlreadyInitialized }', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.9.1/src/registry.rs:170:10
#0 45.01
#0 45.01 Stack:
#0 45.01
#0 45.01 Error
#0 45.01     at module.exports.__wbg_new_693216e109162396 (/app/node_modules/next/wasm/@next/swc-wasm-nodejs/wasm.js:202:17)
#0 45.01     at wasm://wasm/059aa236:wasm-function[5445]:0xf90917
#0 45.01     at wasm://wasm/059aa236:wasm-function[13700]:0x11eca7b
#0 45.01     at wasm://wasm/059aa236:wasm-function[9995]:0x1186adc
#0 45.01     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 45.01     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 45.01     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 45.01     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 45.01     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 45.01     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 45.01
#0 45.01
#0 45.59 Failed to compile.
#0 45.59
#0 45.59 static/chunks/pages/_app-8407aef237a8638f.js from Terser
#0 45.59 unreachable
#0 45.59 RuntimeError: unreachable
#0 45.59     at wasm://wasm/059aa236:wasm-function[9995]:0x1186aff
#0 45.59     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 45.59     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 45.59     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 45.59     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x760098
#0 45.59     at wasm://wasm/059aa236:wasm-function[3154]:0xd6fe73
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f24f
#0 45.59
#0 45.59 static/chunks/pages/cart-30098418c1e82897.js from Terser
#0 45.59 unreachable
#0 45.59 RuntimeError: unreachable
#0 45.59     at wasm://wasm/059aa236:wasm-function[9995]:0x1186afd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 45.59     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 45.59     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 45.59     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x760098
#0 45.59     at wasm://wasm/059aa236:wasm-function[3154]:0xd6fe73
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f24f
#0 45.59
#0 45.59 static/chunks/framework-186861825da4f5ee.js from Terser
#0 45.59 unreachable
#0 45.59 RuntimeError: unreachable
#0 45.59     at wasm://wasm/059aa236:wasm-function[9995]:0x1186afd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 45.59     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 45.59     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 45.59     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x760098
#0 45.59     at wasm://wasm/059aa236:wasm-function[3154]:0xd6fe73
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f24f
#0 45.59
#0 45.59 static/chunks/764-bf797c6df7c4f464.js from Terser
#0 45.59 unreachable
#0 45.59 RuntimeError: unreachable
#0 45.59     at wasm://wasm/059aa236:wasm-function[9995]:0x1186afd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11009]:0x11b6862
#0 45.59     at wasm://wasm/059aa236:wasm-function[12785]:0x11e5cdd
#0 45.59     at wasm://wasm/059aa236:wasm-function[11383]:0x11c358f
#0 45.59     at wasm://wasm/059aa236:wasm-function[10017]:0x1187e72
#0 45.59     at wasm://wasm/059aa236:wasm-function[1151]:0x98decd
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f3f2
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x760098
#0 45.59     at wasm://wasm/059aa236:wasm-function[3154]:0xd6fe73
#0 45.59     at wasm://wasm/059aa236:wasm-function[654]:0x75f24f
#0 45.59
#0 45.59
#0 45.60 > Build failed because of webpack errors
------
failed to solve: executor failed running [/bin/sh -c npm run build]: exit code: 1
```

**Comment by julianocosta89** [comment]
@xoscar would you have some time to take a look at this?
We would love to have that PR merged to do release `1.3.0`.

**Comment by xoscar** [comment]
Hey @julianocosta89 how are you trying to build the image? I did it using `docker compose build frontend --no-cache` and it worked.

@austinlparker I wonder if I should update the doc matrix, the current state of it seems to be quite outdated.

**Review by julianocosta89 — APPROVED** [review]
LGTM, just missing the suggestion from @pichlermarc

**Comment by cartersocha** [comment]
@puckpuck / @open-telemetry/demo-approvers everything addressed? Im on mobile until tonight but I think I see everything resolved

**Comment by puckpuck** [comment]
I will test this all end-end in a few hours, and report back.

**Comment by austinlparker** [comment]
Thanks @puckpuck ping me on slack when it's good

**Review by puckpuck — APPROVED** [review]

**Pull request merged**